python/mlx/optimizers.py

```diff
@@ -221,3 +221,44 @@ def apply_single(
         return super().apply_single(
             gradient, parameter * (1 - self.learning_rate * self.weight_decay), state
         )
+
+
+class Adagrad(Optimizer):
+    r"""Implementation of the Adagrad optimizer [1].
+
+    Our Adagrad implementation follows the original paper and omits the bias
+    correction in the first and second moment estimates. In detail,
+
+    .. math::
+
+        v_{t+1} &= v_t + g_t^2 \\
+        w_{t+1} &= w_t - \lambda \frac{g_t}{\sqrt{v_{t+1} + \epsilon}}
+
+    [1]: Duchi, J., Hazan, E. and Singer, Y., 2011. Adaptive subgradient methods
+    for online learning and stochastic optimization. JMLR 2011.
+    """
+
+    def __init__(self, learning_rate: float, eps: float = 1e-8):
+        super().__init__()
+
+        self.learning_rate = learning_rate
+        self.eps = eps
+
+        if self.learning_rate < 0.0:
+            raise ValueError("Invalid learning rate: {}".format(self.learning_rate))
+        if self.eps < 0.0:
+            raise ValueError("Invalid eps: {}".format(self.eps))
```

Comment:
Maybe too nitpicky but we use format strings almost everywhere in the library for things like that. I think it is slightly more readable as well. For instance,

```python
raise ValueError(f"Adagrad epsilon should be >0, {self.eps} was provided instead")
```

Reply:
done!